File: frontend/stores/room.ts
import { defineStore } from 'pinia'
import { useProfileStore } from '~/stores/profile'

export type RoomStatus = 'in_lobby' | 'in_game' | 'closed'
export type PlayerRole = 'master' | 'player' | 'spectator'

export interface RoomPlayer {
  id: string
  nickname: string
  mascot: string
  color: string
  role: PlayerRole
  is_master: boolean
  is_spectator: boolean
  is_muted: boolean
  is_connected: boolean
  score: number
}

export interface RoomSettings {
  game_type: string
  word_length: number
  round_duration: number
  max_rounds: number
  max_attempts: number
  language: string
}

export interface Room {
  code: string
  status: RoomStatus
  master_id: string
  settings: RoomSettings
  current_round: number
  ends_at?: string
  players: RoomPlayer[]
}

export interface ChatMessage {
  id: string
  sender_id: string
  sender: string
  mascot: string
  color: string
  content: string
  is_system: boolean
  created_at: string
}

export const useRoomStore = defineStore('room', {
  state: () => ({
    currentRoom: null as Room | null,
    chatMessages: [] as ChatMessage[],
  }),

  getters: {
    players: (state) => state.currentRoom?.players || [],
    activePlayers: (state) => (state.currentRoom?.players || []).filter(p => !p.is_spectator),
    spectators: (state) => (state.currentRoom?.players || []).filter(p => p.is_spectator),
    
    isMaster: (state) => {
      const profile = useProfileStore()
      const me = state.currentRoom?.players.find(p => p.nickname === profile.nickname)
      return !!me?.is_master
    },

    me: (state) => {
      const profile = useProfileStore()
      return state.currentRoom?.players.find(p => p.nickname === profile.nickname)
    },

    masterPlayer: (state) => state.currentRoom?.players.find(p => p.is_master),
  },

  actions: {
    setRoom(room: Room) {
      this.currentRoom = room
    },

    addChatMessage(msg: ChatMessage) {
      this.chatMessages.push(msg)
      if (this.chatMessages.length > 50) {
        this.chatMessages.shift()
      }
    },

    setChatHistory(messages: ChatMessage[]) {
      this.chatMessages = messages
    },

    clearRoom() {
      this.currentRoom = null
      this.chatMessages = []
    },
  },
})

File: backend/internal/domain/room.go
package domain

import (
	"encoding/json"
	"time"

	"github.com/google/uuid"
)

type RoomStatus string

const (
	RoomStatusInLobby RoomStatus = "in_lobby"
	RoomStatusInGame  RoomStatus = "in_game"
	RoomStatusClosed  RoomStatus = "closed"
)

type PlayerRole string

const (
	RoleMaster    PlayerRole = "master"
	RolePlayer    PlayerRole = "player"
	RoleSpectator PlayerRole = "spectator"
)

// RoomPlayer représente un joueur dans une salle
type RoomPlayer struct {
	ID          uuid.UUID  `json:"id"`
	Nickname    string     `json:"nickname"`
	Mascot      string     `json:"mascot"` // "dice", "domino", "card", "knight", "d20", "meeple"
	Color       string     `json:"color"`  // Hex code (ex: #FFD300, #1D4ED8, etc.)
	Role        PlayerRole `json:"role"`
	IsMaster    bool       `json:"is_master"`
	IsSpectator bool       `json:"is_spectator"`
	IsMuted     bool       `json:"is_muted"`
	IsConnected bool       `json:"is_connected"`
	JoinedAt    time.Time  `json:"joined_at"`
	LastSeenAt  time.Time  `json:"last_seen_at"`
	Score       int        `json:"score"`
}

// RoomSettings configure la salle de jeu
type RoomSettings struct {
	GameType      string `json:"game_type"`      // "wordle"
	WordLength    int    `json:"word_length"`    // 3 à 8 (défaut 5)
	RoundDuration int    `json:"round_duration"` // secondes (défaut 60)
	MaxRounds     int    `json:"max_rounds"`     // manches (défaut 3)
	MaxAttempts   int    `json:"max_attempts"`   // essais max (défaut 6)
	Language      string `json:"language"`       // "fr"
}

// Room représente l'état global d'une salle
type Room struct {
	Code         string       `json:"code"`
	Status       RoomStatus   `json:"status"`
	MasterID     uuid.UUID    `json:"master_id"`
	Settings     RoomSettings `json:"settings"`
	CurrentRound int          `json:"current_round"`
	SecretWord   string       `json:"secret_word,omitempty"` // Masqué pour les clients en cours de jeu
	EndsAt       *time.Time   `json:"ends_at,omitempty"`
	CreatedAt    time.Time    `json:"created_at"`
	Players      []RoomPlayer `json:"players"`
}

// SessionData stockée dans Redis sous session:{token} avec TTL 45s pour reconnexion transparente
type SessionData struct {
	UserID   uuid.UUID `json:"user_id"`
	RoomCode string    `json:"room_code"`
	Nickname string    `json:"nickname"`
	Mascot   string    `json:"mascot"`
	Color    string    `json:"color"`
}

// ChatMessage représente un message de discussion ou système
type ChatMessage struct {
	ID        string    `json:"id"`
	SenderID  uuid.UUID `json:"sender_id"`
	Sender    string    `json:"sender"`
	Mascot    string    `json:"mascot"`
	Color     string    `json:"color"`
	Content   string    `json:"content"`
	IsSystem  bool      `json:"is_system"`
	CreatedAt time.Time `json:"created_at"`
}

// RoundHistory récapitule une manche passée
type RoundHistory struct {
	RoundNum   int                    `json:"round_num"`
	SecretWord string                 `json:"secret_word"`
	WinnerID   *uuid.UUID             `json:"winner_id,omitempty"`
	WinnerName string                 `json:"winner_name,omitempty"`
	Scores     map[string]int         `json:"scores"` // UserID string -> points de la manche
	EndedAt    time.Time              `json:"ended_at"`
}

// WSMessage format standard de message WebSocket
type WSMessage struct {
	Type    string          `json:"type"`
	Payload json.RawMessage `json:"payload"`
}
